File: analisis_numerico/metodos/algoritmos/spline_cubico.py
import numpy as np
import matplotlib.pyplot as plt
import io
import base64

def spline_cubico(x, y):
    """
    spline_cubico: Interpolación spline cúbico natural.
    Entrada:
        x: nodos (vector de abscisas, ordenado y sin repetidos)
        y: valores en los nodos (vector de ordenadas)
    Salida:
        coef: matriz (n-1)x4 con los coeficientes de cada tramo: [a, b, c, d]
    """
    x = np.asarray(x)
    y = np.asarray(y)

    # Validaciones
    if x.ndim != 1 or y.ndim != 1:
        raise ValueError('Las entradas deben ser vectores.')
    if len(x) != len(y):
        raise ValueError('Los vectores x e y deben tener la misma longitud.')
    if len(np.unique(x)) != len(x):
        raise ValueError('Los valores de x deben ser distintos.')
    if not np.all(np.diff(x) > 0):
        raise ValueError('x debe estar ordenado estrictamente de forma creciente.')

    n = len(x)
    N = 4 * (n - 1)
    A = np.zeros((N, N))
    b_vec = np.zeros(N)

    # Ecuaciones de coincidencia
    row = 0
    col = 0
    for i in range(n - 1):
        xi = x[i]
        A[row, col:col+4] = [xi**3, xi**2, xi, 1]
        b_vec[row] = y[i]
        row += 1
        col += 4

    col = 0
    for i in range(1, n):
        xi = x[i]
        A[row, col:col+4] = [xi**3, xi**2, xi, 1]
        b_vec[row] = y[i]
        row += 1
        col += 4

    # Primera derivada continua en puntos interiores
    col = 0
    for i in range(1, n - 1):
        xi = x[i]
        A[row, col:col+3] = [3*xi**2, 2*xi, 1]
        A[row, col+4:col+7] = [-3*xi**2, -2*xi, -1]
        row += 1
        col += 4

    # Segunda derivada continua en puntos interiores
    col = 0
    for i in range(1, n - 1):
        xi = x[i]
        A[row, col:col+2] = [6*xi, 2]
        A[row, col+4:col+6] = [-6*xi, -2]
        row += 1
        col += 4

    # Condiciones de spline natural
    A[row, 0:2] = [6*x[0], 2]
    row += 1
    A[row, -4:-2] = [6*x[-1], 2]
    row += 1

    # Resolver el sistema
    sol = np.linalg.solve(A, b_vec)
    coef = sol.reshape((n - 1, 4))

    # Mostrar polinomios por tramos
    print("Polinomio Spline Cúbico por tramos:")
    for i in range(n - 1):
        a, b_, c, d = coef[i]
        print(f"P{i+1}(x) = {a:.6f}*x^3 + {b_:.6f}*x^2 + {c:.6f}*x + {d:.6f},  x ∈ [{x[i]:.2f}, {x[i+1]:.2f}]")

    # Graficar
    plt.figure()
    for i in range(n - 1):
        a, b_, c, d = coef[i]
        t = np.linspace(x[i], x[i+1], 100)
        f = a*t**3 + b_*t**2 + c*t + d
        plt.plot(t, f, label=f'Tramo {i+1}')
    plt.plot(x, y, 'ro', label='Puntos de datos')
    plt.grid(True)
    plt.xlabel('x')
    plt.ylabel('Spline Cúbico')
    plt.title('Interpolación Spline Cúbico Natural')
    plt.legend()
    buf = io.BytesIO()
    plt.savefig(buf, format='png')
    plt.close()
    buf.seek(0)
    grafica_base64 = base64.b64encode(buf.read()).decode('ascii')
    return {'coef': coef.tolist(), 'grafica_base64': grafica_base64}

File: SplineCubico.m
function [coef] = SplineCubico(x, y)
    % SplineCubico: Interpolación spline cúbico natural.
    % Entrada:
    %   x: nodos (vector de abscisas, ordenado y sin repetidos)
    %   y: valores en los nodos (vector de ordenadas)
    % Salida:
    %   coef: matriz (n-1)x4 con los coeficientes de cada tramo: [a, b, c, d]

    % Validaciones
    if ~isvector(x) || ~isvector(y)
        error('Las entradas deben ser vectores.');
    end
    if length(x) ~= length(y)
        error('Los vectores x e y deben tener la misma longitud.');
    end
    if length(unique(x)) ~= length(x)
        error('Los valores de x deben ser distintos.');
    end
    if any(diff(x) <= 0)
        error('x debe estar ordenado estrictamente de forma creciente.');
    end

    n = length(x);
    N = 4 * (n - 1);  % Número de incógnitas
    A = zeros(N);
    b = zeros(N, 1);

    % Ecuaciones: coincidencia en los extremos de cada intervalo
    row = 1;
    col = 1;
    for i = 1:n-1
        xi = x(i);
        A(row, col:col+3) = [xi^3, xi^2, xi, 1];
        b(row) = y(i);
        row = row + 1;
        col = col + 4;
    end

    col = 1;
    for i = 2:n
        xi = x(i);
        A(row, col:col+3) = [xi^3, xi^2, xi, 1];
        b(row) = y(i);
        row = row + 1;
        col = col + 4;
    end

    % Derivadas de primer orden iguales en puntos interiores
    col = 1;
    for i = 2:n-1
        xi = x(i);
        A(row, col:col+2) = [3*xi^2, 2*xi, 1];
        A(row, col+4:col+6) = [-3*xi^2, -2*xi, -1];
        b(row) = 0;
        row = row + 1;
        col = col + 4;
    end

    % Derivadas de segundo orden iguales en puntos interiores
    col = 1;
    for i = 2:n-1
        xi = x(i);
        A(row, col:col+1) = [6*xi, 2];
        A(row, col+4:col+5) = [-6*xi, -2];
        b(row) = 0;
        row = row + 1;
        col = col + 4;
    end

    % Condición de spline natural (segunda derivada = 0 en extremos)
    A(row, 1:2) = [6*x(1), 2];  % en x1
    b(row) = 0;
    row = row + 1;
    A(row, end-3:end-2) = [6*x(end), 2];  % en xn
    b(row) = 0;

    % Resolución del sistema
    sol = A \ b;
    coef = reshape(sol, 4, n-1)';  % Cada fila: [a b c d] de a*x^3 + b*x^2 + c*x + d

    % Mostrar los polinomios por tramos
    fprintf('Polinomio Spline Cúbico por tramos:\n');
    for i = 1:n-1
        fprintf('P%d(x) = %.6f*x^3 + %.6f*x^2 + %.6f*x + %.6f,  x ∈ [%.2f, %.2f]\n', ...
            i, coef(i,1), coef(i,2), coef(i,3), coef(i,4), x(i), x(i+1));
    end

    % Graficar los tramos
    figure;
    hold on;
    for i = 1:n-1
        a = coef(i,1); b_ = coef(i,2); c = coef(i,3); d = coef(i,4);
        f = @(t) a*t.^3 + b_*t.^2 + c*t + d;
        fplot(f, [x(i), x(i+1)], 'LineWidth', 2);
    end
    plot(x, y, 'ro', 'MarkerFaceColor', 'r');
    grid on;
    xlabel('x');
    ylabel('Spline Cúbico');
    title('Interpolación Spline Cúbico Natural');
    legend('Tramos cúbicos', 'Puntos de datos', 'Location', 'best');
    hold off;
end
